test/tests.zig

```diff
@@ -772,6 +772,47 @@ pub fn addCliTests(b: *std.Build) *Step {
         step.dependOn(&cleanup.step);
     }
 
+    // Test `zig cc -x c -`
+    // Test author was not able to figure out how to start a child process and
+    // give an fd to it's stdin. fork/exec works, but we are limiting ourselves
+    // to POSIX.
+    if (builtin.os.tag != .windows) {
+        const tmp_path = b.makeTempPath();
+        var dir = std.fs.cwd().openDir(tmp_path, .{}) catch @panic("unhandled");
+        dir.writeFile("truth.c", "int main() { return 42; }") catch @panic("unhandled");
+        var infile = dir.openFile("truth.c", .{}) catch @panic("unhandled");
+
+        const outfile = std.fs.path.joinZ(
+            b.allocator,
+            &[_][]const u8{ tmp_path, "truth" },
+        ) catch @panic("unhandled");
+        const pid_result = std.os.fork() catch @panic("unhandled");
+        if (pid_result == 0) { // child
+            std.os.dup2(infile.handle, std.os.STDIN_FILENO) catch @panic("unhandled");
+            const argv = &[_:null]?[*:0]const u8{
+                b.zig_exe, "cc",
+                "-o",      outfile,
+                "-x",      "c",
+                "-",
+            };
+            const envp = &[_:null]?[*:0]const u8{
+                std.fmt.allocPrintZ(b.allocator, "ZIG_GLOBAL_CACHE_DIR={s}", .{tmp_path}) catch @panic("unhandled"),
+            };
+            const err = std.os.execveZ(b.zig_exe, argv, envp);
+            std.debug.print("execve error: {any}\n", .{err});
+            std.os.exit(1);
+        }
+
+        const res = std.os.waitpid(pid_result, 0);
+        assert(0 == res.status);
+
+        // run the compiled executable and check if it's telling the truth.
+        _ = exec(b.allocator, tmp_path, 42, &[_][]const u8{outfile}) catch @panic("unhandled");
+
+        const cleanup = b.addRemoveDirTree(tmp_path);
+        step.dependOn(&cleanup.step);
+    }
```

Comment:
This test as is does the following things directly in the configure phase rather than using build steps like it must do in order to play nicely with the other build steps:
 * making a temp dir and writing a file
 * calling fork, execve, and wait
 * calling exec

If you look at the Test Godbolt API code block above, it shows how to use the build system API for testing. It will be a combination of WriteFileStep, and RunStep.

I'm happy to provide assistance in writing this test.